<a href="https://colab.research.google.com/github/kbezzy4/stc510/blob/main/Module_3_Basics_Project_Bell.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [155]:
# Importing the needed modules
import pandas as pd
import numpy as np
import datetime as dt

In [157]:
# Pulling in the data
data = pd.read_csv('crimedata.csv')

<ipython-input-157-ec18562d66d5>:2: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv('crimedata.csv')


In [158]:
# Making sense of the data by getting info on columns:
df = pd.DataFrame(data)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 524115 entries, 0 to 524114
Data columns (total 9 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   _id                 524115 non-null  int64  
 1   INC NUMBER          524115 non-null  object 
 2   OCCURRED ON         523696 non-null  object 
 3   OCCURRED TO         337814 non-null  object 
 4   UCR CRIME CATEGORY  524115 non-null  object 
 5   100 BLOCK ADDR      524115 non-null  object 
 6   ZIP                 524109 non-null  float64
 7   PREMISE TYPE        520633 non-null  object 
 8   GRID                522682 non-null  object 
dtypes: float64(1), int64(1), object(7)
memory usage: 36.0+ MB


In [159]:
# Renaming columns because of spacing in order to reference easier and confirming the renaming:
df.rename(columns={'INC NUMBER': 'INC_NUMBER'}, inplace=True)
df.rename(columns={'OCCURRED ON': 'OCCURRED_ON'}, inplace=True)
df.rename(columns={'OCCURRED TO': 'OCCURRED_TO'}, inplace=True)
df.rename(columns={'UCR CRIME CATEGORY': 'UCR_CRIME_CATEGORY'}, inplace=True)
df.rename(columns={'100 BLOCK ADDR': '100_BLOCK_ADDR'}, inplace=True)
df.rename(columns={'PREMISE TYPE': 'PREMISE_TYPE'}, inplace=True)
df

_id       INC_NUMBER        OCCURRED_ON        OCCURRED_TO  \
0            1  201500002101405  11/01/2015  00:00  11/01/2015  05:00   
1            2  201600000052855  11/01/2015  00:00  01/09/2016  00:00   
2            3  201500002102327  11/01/2015  00:00  11/01/2015  09:00   
3            4  201500002102668  11/01/2015  00:00  11/01/2015  11:50   
4            5  201600000527709  11/01/2015  00:00  03/22/2016  00:36   
...        ...              ...                ...                ...   
524110  524111  202400000116889  01/23/2024  21:51  01/23/2024  22:04   
524111  524112  202400000113080  01/23/2024  22:23  01/23/2024  22:23   
524112  524113  202400000113191  01/23/2024  23:08                NaN   
524113  524114  202400000113227  01/23/2024  23:16  01/23/2024  23:16   
524114  524115  202400000116334  01/23/2024  23:53  01/24/2024  00:00   

         UCR_CRIME_CATEGORY                   100_BLOCK_ADDR      ZIP  \
0       MOTOR VEHICLE THEFT              102XX W MEDLOCK AVE  85307.0   
1       MOTOR VEHICLE THEFT         N 43RD AVE & W CACTUS RD  85029.0   
2             LARCENY-THEFT                   51XX N 15TH ST  85014.0   
3       MOTOR VEHICLE THEFT                   69XX W WOOD ST  85043.0   
4             LARCENY-THEFT              33XX W CAMELBACK RD  85017.0   
...                     ...                              ...      ...   
524110        LARCENY-THEFT  126XX N PARADISE VILLAGE PKWY W  85032.0   
524111   AGGRAVATED ASSAULT              26XX W GLENROSA AVE  85017.0   
524112         DRUG OFFENSE  N 43RD AVE & W INDIAN SCHOOL RD  85031.0   
524113         DRUG OFFENSE              25XX W CAMELBACK RD  85017.0   
524114  MOTOR VEHICLE THEFT                   22XX E YALE ST  85006.0   

                       PREMISE_TYPE   GRID  
0               SINGLE FAMILY HOUSE   BJ04  
1               SINGLE FAMILY HOUSE   DA19  
2                         APARTMENT   BJ30  
3               SINGLE FAMILY HOUSE   AF12  
4                       PARKING LOT   BJ21  
...                             ...    ...  
524110              RETAIL BUSINESS   DB37  
524111              RETAIL BUSINESS  BH23B  
524112  STREET / ROADWAY / SIDEWALK   BH18  
524113                      VEHICLE  BI23B  
524114  STREET / ROADWAY / SIDEWALK   BE32  

[524115 rows x 9 columns]

In [160]:
# Getting some value counts for the crime category
df.UCR_CRIME_CATEGORY.value_counts()

LARCENY-THEFT                            266654
BURGLARY                                  67379
MOTOR VEHICLE THEFT                       60029
DRUG OFFENSE                              47677
AGGRAVATED ASSAULT                        47484
ROBBERY                                   20635
RAPE                                       8075
ARSON                                      4837
MURDER AND NON-NEGLIGENT MANSLAUGHTER      1345
Name: UCR_CRIME_CATEGORY, dtype: int64

In [161]:
# To find crime rate in 2023, only have crimes in 2023; converting "occurred on" to datetime and filtering
df['OCCURRED_ON'] = pd.to_datetime(df['OCCURRED_ON'])
res = df.loc[(df['OCCURRED_ON'] >= '2023-01-01') & (df['OCCURRED_ON']<='2024-01-01')]
print(res)

           _id       INC_NUMBER OCCURRED_ON        OCCURRED_TO  \
465199  465200  202300001630322  2023-01-01  09/10/2023  00:00   
465200  465201  202300000014949  2023-01-01  01/03/2023  23:20   
465201  465202  202300001254671  2023-01-01                NaN   
465202  465203  202300001184679  2023-01-01  01/31/2023  23:59   
465203  465204  202300001030551  2023-01-01  03/18/2023  00:00   
...        ...              ...         ...                ...   
520876  520877  202400000023908  2024-01-01  01/01/2024  00:00   
520877  520878  202400000005258  2024-01-01  01/01/2024  23:31   
520878  520879  202400000017366  2024-01-01  01/01/2024  23:00   
520879  520880  202400000017323  2024-01-01  01/04/2024  12:21   
520880  520881  202400000003538  2024-01-01                NaN   

         UCR_CRIME_CATEGORY               100_BLOCK_ADDR      ZIP  \
465199        LARCENY-THEFT               6XX S 19TH AVE  85007.0   
465200  MOTOR VEHICLE THEFT          44XX W ENCANTO BLVD  85035.0   


In [162]:
# Getting value counts for crimes that year
res.UCR_CRIME_CATEGORY.value_counts()

LARCENY-THEFT                            25705
MOTOR VEHICLE THEFT                       8513
AGGRAVATED ASSAULT                        6518
DRUG OFFENSE                              6162
BURGLARY                                  5213
ROBBERY                                   1807
RAPE                                       884
ARSON                                      693
MURDER AND NON-NEGLIGENT MANSLAUGHTER      187
Name: UCR_CRIME_CATEGORY, dtype: int64

In [163]:
# Population data needed for crime rate; population in 2023: 7,431,344
pop_az_23 = 7431344
x1 = np.array([25705,8513,6518,6162,5213,1807,884,693,187])
crime_rate_23 = ((x1/pop_az_23)*100000)
crime_rate_23

array([345.89974573, 114.55532135,  87.70957178,  82.91905206,
        70.1488183 ,  24.31592455,  11.89556021,   9.32536564,
         2.51636851])

In [165]:
#Repeating last three lines of code, but for year 2018. AZ pop. in 2018: 7,164,000
df['OCCURRED_ON'] = pd.to_datetime(df['OCCURRED_ON'])
res2 = df.loc[(df['OCCURRED_ON'] >= '2018-01-01') & (df['OCCURRED_ON']<='2019-01-01')]
print(res2)

           _id       INC_NUMBER OCCURRED_ON        OCCURRED_TO  \
139796  139797  201900002265971  2018-01-01  12/19/2019  00:00   
139797  139798  201800000099561  2018-01-01  01/05/2018  23:59   
139798  139799  202000000443619  2018-01-01  01/01/2018  00:00   
139799  139800  202000000794824  2018-01-01  12/31/2018  23:59   
139800  139801  202000000953525  2018-01-01  12/24/2018  00:00   
...        ...              ...         ...                ...   
204865  204866  201900000164215  2019-01-01  01/29/2019  00:00   
204866  204867  201900000409390  2019-01-01  03/10/2019  16:16   
204867  204868  201900000332070  2019-01-01  02/26/2019  00:00   
204868  204869  201900000525136  2019-01-01  03/29/2019  00:00   
204869  204870  201900000352448  2019-01-01  03/01/2019  00:00   

        UCR_CRIME_CATEGORY       100_BLOCK_ADDR      ZIP  \
139796       LARCENY-THEFT       34XX E WOOD ST  85040.0   
139797            BURGLARY      188XX N 35TH PL  85050.0   
139798                RAPE 

In [166]:
res2.UCR_CRIME_CATEGORY.value_counts()

LARCENY-THEFT                            33203
BURGLARY                                  9303
MOTOR VEHICLE THEFT                       6910
DRUG OFFENSE                              5734
AGGRAVATED ASSAULT                        5393
ROBBERY                                   2820
RAPE                                      1049
ARSON                                      460
MURDER AND NON-NEGLIGENT MANSLAUGHTER      113
Name: UCR_CRIME_CATEGORY, dtype: int64

In [168]:
pop_az_18 = 7164000
x2 = np.array([33203,9303,6910,5734,5393,2820,1049,460,113])
crime_rate_18 = ((x2/pop_az_18)*100000)
crime_rate_18

array([463.47012842, 129.85762144,  96.4544947 ,  80.03908431,
        75.27917365,  39.36348409,  14.64265773,   6.42099386,
         1.5773311 ])

In [169]:
# Getting percent change in crime rate from 2018 to 2023
crime_rate_diff = ((crime_rate_23)-(crime_rate_18)/(abs(crime_rate_18))*100)
crime_rate_diff

array([245.89974573,  14.55532135, -12.29042822, -17.08094794,
       -29.8511817 , -75.68407545, -88.10443979, -90.67463436,
       -97.48363149])



---


LARCENY-THEFT had the greatest percent change from 2018 - 2023 while MURDER and NON-NEGLIGENT MANSLAUGHTER had the lowest percent change from 2018 - 2023

In [170]:
# Sort the dataframe by crime category and zip code
df.sort_values(by=['UCR_CRIME_CATEGORY','ZIP'])

_id       INC_NUMBER         OCCURRED_ON        OCCURRED_TO  \
537        538  201500002122070 2015-11-04 08:47:00                NaN   
1879      1880  201500002177147 2015-11-12 15:23:00  11/12/2015  15:25   
5737      5738  201500002325071 2015-12-05 23:15:00  12/05/2015  23:21   
9028      9029  201500002441181 2015-12-24 15:27:00                NaN   
9766      9767  201510000020034 2015-12-29 16:55:00                NaN   
...        ...              ...                 ...                ...   
522795  522796  202400000064129 2024-01-14 01:39:00                NaN   
468157  468158  202300000094663 2023-01-19 16:50:00  01/19/2023  16:51   
137904  137905  201700002229242 2017-12-20 09:56:00                NaN   
156946  156947  201800000608354 2018-04-07 02:00:00  04/07/2018  02:15   
244108  244109  201900001289065 2019-07-28 12:08:00                NaN   

        UCR_CRIME_CATEGORY                 100_BLOCK_ADDR      ZIP  \
537     AGGRAVATED ASSAULT                4XX W MONROE ST  85003.0   
1879    AGGRAVATED ASSAULT       S 1ST AVE & W MADISON ST  85003.0   
5737    AGGRAVATED ASSAULT               XX N CENTRAL AVE  85003.0   
9028    AGGRAVATED ASSAULT  N CENTRAL AVE & W MCDOWELL RD  85003.0   
9766    AGGRAVATED ASSAULT             2XX W VAN BUREN ST  85003.0   
...                    ...                            ...      ...   
522795             ROBBERY                29XX S 99TH AVE  85353.0   
468157             ROBBERY               248XX N 72ND AVE  85383.0   
137904             ROBBERY               58XX W THOMAS RD  85396.0   
156946             ROBBERY               30XX W THOMAS RD  85396.0   
244108             ROBBERY               65XX W THOMAS RD  85396.0   

                               PREMISE_TYPE  GRID  
537                                     NaN  BA27  
1879    STREET / ROADWAY / ALLEY / SIDEWALK  BA27  
5737                            PARKING LOT  BA27  
9028                             LIGHT RAIL  BD27  
9766    STREET / ROADWAY / ALLEY / SIDEWALK  BB27  
...                                     ...   ...  
522795           CONVENIENCE MARKET / STORE  AG05  
468157          STREET / ROADWAY / SIDEWALK   NaN  
137904                      RETAIL BUSINESS   NaN  
156946                          PARKING LOT   NaN  
244108                              VEHICLE   NaN  

[524115 rows x 9 columns]

In [171]:
#Relative Frequencies for different crimes using lambda function
df.groupby('UCR_CRIME_CATEGORY').apply(lambda x: x['UCR_CRIME_CATEGORY'].count()/df.shape[0])

UCR_CRIME_CATEGORY
AGGRAVATED ASSAULT                       0.090598
ARSON                                    0.009229
BURGLARY                                 0.128558
DRUG OFFENSE                             0.090967
LARCENY-THEFT                            0.508770
MOTOR VEHICLE THEFT                      0.114534
MURDER AND NON-NEGLIGENT MANSLAUGHTER    0.002566
RAPE                                     0.015407
ROBBERY                                  0.039371
dtype: float64



---


**Larceny-theft is most common over all.**

In [172]:
# Grouping by UCR and ZIP
df_zip = df.groupby(['UCR_CRIME_CATEGORY','ZIP']).size()
df_zip

UCR_CRIME_CATEGORY  ZIP    
AGGRAVATED ASSAULT  85003.0     720
                    85004.0     635
                    85006.0    1196
                    85007.0    1393
                    85008.0    2427
                               ... 
ROBBERY             85339.0     431
                    85345.0       1
                    85353.0     144
                    85383.0       1
                    85396.0       3
Length: 689, dtype: int64

In [173]:
# Printing df_zip as string
print(df_zip.to_string())

UCR_CRIME_CATEGORY                     ZIP    
AGGRAVATED ASSAULT                     85003.0      720
                                       85004.0      635
                                       85006.0     1196
                                       85007.0     1393
                                       85008.0     2427
                                       85009.0     2415
                                       85012.0      259
                                       85013.0      806
                                       85014.0      716
                                       85015.0     2459
                                       85016.0      878
                                       85017.0     2441
                                       85018.0      646
                                       85019.0     1315
                                       85020.0      997
                                       85021.0     1906
                                       85022.0     1002
 

   

---


*   Aggravated assault and Burglary, Larceny-theft is most common in ZIP 85015
*   Arson & Motor Vehicle Theft is most common in ZIP 85009
*   Burglary is most common in ZIP 85051
*   Drug Offense & Murder/Non-negligle Manslaughter & Robbery is most common in ZIP 85017
*   Rape is most common in 85008  




In [175]:
df.to_csv('crimedevil.csv')